crates/zome_types/src/query.rs

```diff
@@ -21,6 +26,87 @@ pub struct ChainQueryFilter {
+/// An agents chain elements returned from a agent_activity_query
+pub struct AgentActivity {
+    /// Headers on this chain.
+    pub activity: Vec<Activity>,
+    /// The status of this chain.
+    pub status: ChainStatus,
+    /// The highest chain header that has
+    /// been observed by this authority.
+    pub highest_observed: Option<HighestObserved>,
+}
+
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+/// Individual agent activity.
+/// The header with it's validation status.
+// TODO: This is a little weird because when we have warrants
+// we will have a warrant for each position in the chain that
+// is warranted. Maybe that makes this redundant?
+pub struct Activity {
+    /// The header on this chain
+    pub header: SignedHeaderHashed,
+    /// The validation status of this individual header
+    pub validation_status: ValidationStatus,
+}
+
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+/// The highest header sequence observed by this authority.
+/// This also includes the headers at this sequence.
+/// If there is more then one then there is a fork.
+///
+/// This type is to prevent headers being hidden by
+/// withholding the previous header.
+///
+/// The information is tracked at the edge of holochain before
+/// validation (but after drop checks).
+pub struct HighestObserved {
+    /// The highest sequence number observed.
+    pub header_seq: u32,
+    /// Hashes of any headers claiming to be at this
+    /// header sequence.
+    pub hash: Vec<HeaderHash>,
+}
+
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+/// Status of the agent activity chain
+// TODO: In the future we will most likely be replaced
+// by warrants instead of Forked / Invalid so we can provide
+// evidence of why the chain has a status.
+pub enum ChainStatus {
+    /// This authority has no information on the chain.
+    Empty,
+    /// The chain is valid as at this header sequence and header hash.
+    Valid(ChainHead),
+    /// Chain is forked.
+    Forked(ChainFork),
+    /// Chain is invalid because of this header.
+    Invalid(HeaderHash),
+}
+
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+/// The header at the head of the complete chain.
+/// This is as far as this authority can see a
+/// chain with no gaps.
+pub struct ChainHead {
+    /// Sequence number of this chain head.
+    pub header_seq: u32,
+    /// Hash of this chain head
+    pub hash: HeaderHash,
+}
+
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+/// The chain has been forked by these two headers
+pub struct ChainFork {
+    /// The point where the chain has forked.
+    pub fork_seq: u32,
+    /// The first header at this sequence position.
+    pub first_header: HeaderHash,
+    /// The second header at this sequence position.
+    pub second_header: HeaderHash,
```

Comment:
A chain can be forked more than two ways, so should this be a `Vec<HeaderHash>`?

Reply:
See the hackmd but basically `ChainStatus` is about the minimal amount of information to show a chain is valid / invalid. It's not cheap to find forks so it's much better to find and send just the first fork and not have to find all forks in the chain.